### DISCUSSION TITLE
StorageProvider.OpenFolderPickerAsync() not work

### DISCUSSION THREAD

**Discussion by perfect100** [AvaloniaUI/Avalonia#15775] [Q&A]
Hi there,
I'm trying to open a folder picker in avalonia , but the behavior is tricky :

Most Pcs work good, but only one Pc can not open the folder, and the UI blocks, ```OpenFolderPickerAsync()``` never return.

``` C#
public async Task<string?> ShowFolderDialogAsync(string title)
{
    var storageFiles = await StorageProvider.OpenFolderPickerAsync(
        new FolderPickerOpenOptions
        {
            Title = title,
        });

    return storageFiles.Count == 0 ? null : storageFiles[0].Path.LocalPath;
}
```
> This method is called in UI thread.

However, after I tried ugly codes, it looks like that the folder is related to the ```breakpoint``` when debugging(maybe not).

![image](https://github.com/AvaloniaUI/Avalonia/assets/17464894/40190563-4185-4629-bd14-83d13dc30331)

For context, if I put a breakpoint in ```L15```, the folder picker displays, otherwise, UI blocks even I use ```await Task.Delay(2000)```.
So, what happed here? Is the StorageProvider waitting for some condition to meet?

Avalonia : 11.1.0-beta1
OS : Windows 10

**Comment by thevortexcloud** [comment]
This code in the picture looks odd to me.
* Why are you running a background task for something that is already async? 
* Why are you double awaiting? 
* What is `App.GetCurrentActiveWindow` doing? 

Also: 
* How are you getting the result of `ShowFolderDialogAsync`? You may have a deadlock somewhere. 
* Where are you running this code? You should not be trying to access any view related functionality in your view models, that breaks MVVM and has been known to cause weird issues.
* Tasks are not guaranteed to return on the same thread that started them, do you have any synchronisation stuff (EG mutexes)?

**Reply by perfect100** [reply]
Hi @thevortexcloud, thanks for your reply. I'm sorry for make a mistake in the code of the picture.  
```App.GetCurrentActiveWindow``` is actually the ```StorageProvider``` in the code above. And this is the implementation:
```C#
public Window GetCurrentActiveWindow => 
    (ClassicDesktopStyleApplicationLifetime)((Application.Current as App).ApplicationLifetime).Windows.FirstOrDefault(n => n.IsActive)
```

> Why are you running a background task for something that is already async?
> Why are you double awaiting?

I'm just try to verify that does it related to the thread, so I create a new Task. The first code snippet would run in UI thread, which cannot work.

> How are you getting the result of ShowFolderDialogAsync? 

Just ignore the picture, focus on the code snippet. ```await StorageProvider.OpenFolderPickerAsync(...)``` funciton never return.

> Where are you running this code?

I use the ```CommunityToolkit.Mvvm``` from Microsoft, and create a ```RelayCommand```, then bind the command.

```C#
[RelayCommand]
private async Task SetProjectFolderAsync()
{
    var selectedFolder = await this.FileDialogService.ShowFolderDialogAsync("title");
    if (!string.IsNullOrWhiteSpace(selectedFolder))
    {
        // ...
    }
}
```

For your information, only one machine cannot open the folder if run the code snippet. Actually I havn't found the root cause yet.

**Reply by thevortexcloud** [reply]
If I am following your code properly: You are opening the file dialogue from your view model almost directly (via the FileDialogService)?  As such have you: 

* Tried opening it from the code behind of your view?
* Tried forcing it onto the UI thread using a dispatcher call? You are not guaranteed to actually be on the UI thread in this situation

As I mentioned before you can get weird issues trying to directly open dialogues from your view model (not to mention it kind of breaks MVVM). You can try also implementing your dialoge service using one of the patterns in the samples.

https://github.com/AvaloniaUI/Avalonia.Samples?tab=readme-ov-file#view-interaction-samples

I typically just use ReactiveUI's interactions and I don't have issues.

**Reply by thevortexcloud** [reply]
Also your issue sounds identical to this: https://github.com/AvaloniaUI/Avalonia/discussions/13484

That was resolved by some people by opening the dialogue from the view instead of the view model.

**Reply by perfect100** [reply]
Many thanks. I will have a try. 👍

**Reply by perfect100** [reply]
@thevortexcloud , I have canceled the command binding, and use the button click event directly.
So weird! The folder picker never appears , and UI is blocked.

```C#
// we have a button with name of [button] in the Window control.
private async void Button_Click( ...)
{
    var topLevel = TopLevel.GetTopLevel(this.button);
    if (topLevel != null )
    {
        var selectedFiles = await topLevel.StorageProvider.OpenFolderAsync(
        new FolderPickerOpenOptions { Title = "title", AllowMultiple = false, } );

       Console.WriteLine(selectedFiles);
    }
}
```

**Reply by perfect100** [reply]
By the way, the current Window is a modal dialog window based on the MainWindow.

```C#
var mainWindow = // get main window.
var currentWindow = // the view of the MVVM with a button
await currentWindow.ShowDialog(mainWindow);
```

**Reply by thevortexcloud** [reply]
Does opening a dialogue from within a dialogue work? I have never thought to actually try that since it seems weird to me. Are you sure the window is not actually opening and has not just gone behind another dialogue?

**Reply by perfect100** [reply]
> Does opening a dialogue from within a dialogue work?

Yes, I create a new project. Open a new dialogue from main window, and then open the folder picker from the dialogue, it works.

> Are you sure the window is not actually opening and has not just gone behind another dialogue?

The folder picker never shows up actually. I can not find any dialogue behind.
Anyway, maybe some logic issue in my code, I'm refactoring currently.
Many thanks.

**Reply by perfect100** [reply]
After my investigation, I find that there're some  weird issue when calling FolderPicker.

https://github.com/AvaloniaUI/Avalonia/blob/b30894cb5c1e2403b70e72853ffbd0176ad7c6dd/src/Windows/Avalonia.Win32/Win32StorageProvider.cs#L155

If I debug the code (it means how to set the property of the ```frm```), then the folder will show;
But, I just run the code without debug, then the ```var showResult = frm.Show(_windowImpl.Handle.Handle);``` will never return.

So how I can diagnose the issue? 

> Avalonia Version : 11.1.0-beta1

**Reply by jp2masa** [reply]
Did you try adding the `STAThread` attribute to `Main`? I don't know much about COM, but I think it's worth a try.

**Reply by perfect100** [reply]
It already has the ```STAThread``` attribute there...

**Comment by perfect100** [comment]
The root cause is that the ```Main``` entrance function is ```async```, which broken the ```[STA]```...

I have created a issue here : #15806

Thanks all @thevortexcloud @jp2masa
